### PR TITLE
feat: Detect Vyper version from source

### PR DESCRIPTION
**Pull request by DanielSchiavini**
### What I did
- Add two functions that allow to:
  - Detect the version specifier in the pragma given some source code
  - Detect the best version given the specifier
- Add tests

### How to verify it
- tests are included

### Checklist

- [x] I have confirmed that my PR passes all linting checks
- [x] I have included test cases
- [ ] I have updated the documentation (README.md)
- [x] I have added an entry to the changelog

### CODE REVIEW

File: tests/test_detect_vyper_version.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    "version_str,decorator",
+    [
+        ("0.1.0b17", "public"),
+        ("0.3.0-beta17", "external"),
+        ("0.4.0rc6", "external"),
+    ],
+)
+def test_detect_vyper_version_beta(version_str, decorator):
+    source = f"""
+# @version {version_str}
```
Comment (charles-cooper) [inline tests/test_detect_vyper_version.py:21]: test for new `pragma version ...`?

File: vvm/main.py [hunk 2]
```diff
@@ -57,6 +82,9 @@ def compile_source(
     Dict
         Compiler output. The source file name is given as `<stdin>`.
     """
+    if vyper_version is None:
+        vyper_version = detect_vyper_version_from_source(source)
```
Comment (charles-cooper) [inline vvm/main.py:86]: spooky

Comment (DanielSchiavini) [reply vvm/main.py:86]: It's already in the code, why not?

Comment (DanielSchiavini) [reply vvm/main.py:86]: we could add a `detect_version` kwarg

Comment (charles-cooper) [reply vvm/main.py:86]: yea i mean i think it's convenient, but it's a semantics breaking change, downstream users may(?) somehow depend on the existing behavior

File: vvm/main.py [hunk 3]
```diff
@@ -23,6 +30,65 @@ def get_vyper_version() -> Version:
...
+    """
+    Pick the latest vyper version that is installed and satisfies the given specifier.
+    If None of the installed versions satisfy the specifier, pick the latest installable
+    version.
+
+    Arguments
+    ---------
+    specifier : SpecifierSet
+        Specifier to pick a version for.
+    prereleases : bool, optional
```
[the remark this answers is not in the document]
Comment (DanielSchiavini) [reply vvm/main.py:72]: the `packaging` app will only choose prereleases when they are the only ones available

[the remark this answers is not in the document]
Comment (charles-cooper) [reply vvm/main.py:72]: i'm more talking about the default value -- no need to pick between None, True and False

[the remark this answers is not in the document]
Comment (DanielSchiavini) [reply vvm/main.py:72]: I think None is the correct value here

File: tests/conftest.py [hunk 4]
```diff
@@ -64,12 +64,20 @@ def all_versions(request):
     return version
 
 
+@pytest.fixture
+def latest_version():
+    global VERSIONS
+    return VERSIONS[0]
```
Comment (charles-cooper) [inline tests/conftest.py:70]: stray fixture

File: vvm/utils/versioning.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        Vyper version that satisfies the specifier, or None if no version satisfies the specifier.
+    """
+    versions = itertools.chain(
+        get_installed_vyper_versions() if check_installed else [],
+        get_installable_vyper_versions() if check_installable else [],
+    )
+    try:
+        return next(specifier.filter(versions, prereleases))
+    except StopIteration:
+        raise UnexpectedVersionError(f"No installable Vyper satisfies the specifier {specifier}")
```
Comment (charles-cooper) [inline vvm/utils/versioning.py:83]: with no try/catch blocks
```suggestion
    ret = next(specifier.filter(versions, prereleases), None)
    if ret is not None:
        return ret
    raise UnexpectedVersionError(f"No installable Vyper satisfies the specifier {specifier}")
```

File: vvm/utils/versioning.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        if specifier in ("~", "^"):
+            # convert from npm-style to pypi-style
+            if specifier == "^":
+                # minor match, remove the patch from the version
+                version_str = ".".join(version_str.split(".")[:-1])
+            specifier = "~="  # finds compatible versions
+        specifier = _SPECIFIER_OVERRIDES.get(specifier, specifier) or "=="
+        return Specifier(specifier + version_str)
+    except (StopIteration, InvalidSpecifier):
+        return None
```
Comment (charles-cooper) [inline vvm/utils/versioning.py:43]: also with no try/catch block, hard to follow at first glance where the exceptional cases happen
```suggestion
    match = _VERSION_RE.search(source_code)
    if match is None:
        return None

    specifier, version_str = match.groups()
    if specifier in ("~", "^"):
        # convert from npm-style to pypi-style
        if specifier == "^":
            # minor match, remove the patch from the version
            version_str = ".".join(version_str.split(".")[:-1])
        specifier = "~="  # finds compatible versions
    specifier = _SPECIFIER_OVERRIDES.get(specifier, specifier) or "=="
    try:
        return Specifier(specifier + version_str)
    except InvalidSpecifier:
        return None
```

**Review by charles-cooper — COMMENTED** [review]
looks good, left some small style comments

File: tests/conftest.py (hunk 4 above)
Comment (DanielSchiavini) [reply tests/conftest.py:70]: It's used here: https://github.com/vyperlang/vvm/blob/3aa628797bbbc538dfc4761573c671120087c2f6/tests/test_versioning.py#L45-L46

File: vvm/utils/versioning.py (hunk 5 above)
Comment (DanielSchiavini) [reply vvm/utils/versioning.py:83]: try/catch is a bit worse for performance probably, but I find it a bit easier to read. Takes the same amount of lines. It really doesn't matter much imo

File: vvm/utils/versioning.py (hunk 6 above)
Comment (DanielSchiavini) [reply vvm/utils/versioning.py:43]: Agree with you. It used to make sense when the method was just 2 lines

File: tests/test_versioning.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    specifier = _detect_version_specifier(foo_source)
+    assert str(specifier) == f"=={all_versions}"
+    assert all_versions.major == 0
+    assert _pick_vyper_version(specifier) == all_versions
+
+
+@pytest.mark.parametrize(
+    "version_str,decorator,pragma,expected_specifier,expected_version",
+    [
+        ("^0.1.1", "public", "@version", "~=0.1", "latest"),
```
Comment (DanielSchiavini) [inline tests/test_versioning.py:26]: Something might be wrong here.
According to [the documentation](https://peps.python.org/pep-0440/#compatible-release), `~=0.1` should match `0.1.x` not the latest!

Comment (DanielSchiavini) [reply tests/test_versioning.py:26]: My mistake, `~=0.1` matches `0.x`
Which is the same as the npm `^` which allows minor upgrades (major still `v0`).

File: vvm/install.py [hunk 8]
```diff
@@ -162,6 +163,7 @@ def _get_releases(headers: Optional[Dict]) -> Dict:
     return data.json()
 
 
+@functools.lru_cache(maxsize=1, typed=False)
```
Comment (charles-cooper) [inline vvm/install.py:166]: this can be invalidated though by installing new vyper versions right?

Comment (DanielSchiavini) [reply vvm/install.py:166]: This is not about installed versions but installable versions.
When a new Vyper version is released, the application would need to be restarted, given the cache is in memory.
Otherwise, the user would need to call `get_installable_vyper_versions.cache_clear()`

Comment (charles-cooper) [reply vvm/install.py:166]: in that case why limit the cache size to 1? also, let's add a comment
```suggestion
# When a new Vyper version is released, the application would need to be restarted, given the cache is in memory.
# Otherwise, the user would need to call get_installable_vyper_versions.cache_clear()
@functools.lru_cache
```

Comment (charles-cooper) [reply vvm/install.py:166]: also i'm not sure a function which takes dict as a param works for lru_cache --

https://docs.python.org/3/library/functools.html#functools.lru_cache
> Since a dictionary is used to cache results, the positional and keyword arguments to the function must be [hashable](https://docs.python.org/3/glossary.html#term-hashable).

File: vvm/install.py [hunk 9]
```diff
@@ -180,7 +189,8 @@ def get_installable_vyper_versions(headers: Dict = None) -> List[Version]:
         asset = next((i for i in release["assets"] if _get_os_name() in i["name"]), False)
         if asset:
             version_list.append(version)
-    return sorted(version_list, reverse=True)
+
+    return (_installable_vyper_versions := sorted(version_list, reverse=True))
```
Comment (charles-cooper) [inline vvm/install.py:193]: bit hard to parse since `_installable_vyper_versions` is a global
```suggestion
    _installable_vyper_versions = sorted(version_list, reverse=True)
    return _installable_vyper_versions
```

**Review by charles-cooper — APPROVED** [review]
